### ISSUE TITLE
Enhance fp16 compatibility with SyncBN

### ISSUE THREAD

**Issue by FrankLeeeee** [hpcaitech/ColossalAI#258]
## Feature Request

When using naive amp or zero, we need to convert the whole model to fp16 for better memory management. In data parallel training, we need to convert the vanilla batch normalization layer to sync batch normalization layer so that the running statistics are consistent across different ranks. However, the current fp16 implementation converts the `count` in batch norm to fp16 as well and some internal data type checks will fail. If no sync bn is used, models on different ranks will produce different results even given the same inputs. Thus, we should support sync bn with FP16.

There are several ways for implementation and open for discussion
1. write colossalai's own sync bn
2. adapt fp16 module to support a mixture of fp16 and fp32

## To Reproduce

```python
# create resnet
model = resnet18(num_classes=10).cuda()
model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model, process_group=gpc.get_group(ParallelMode.DATA))
model = model.half()

...

for img, label in dataloader:
      output = model(img)
      ...

```

This will raise exception `RuntimeError: Expected counts to have type Half but got Float`

## Screenshots

<img width="1084" alt="Screenshot 2022-02-24 at 5 58 16 PM" src="https://user-images.githubusercontent.com/31818963/155501826-2b797759-94b5-411a-a88a-0ebc710a9057.png">

**Comment by FrankLeeeee** [comment]
Perhaps, we need to write our own sync BN layer for compatibility.

**Comment by FrankLeeeee** [comment]
Fixed by #291
